### ISSUE TITLE
Lambda local artifact not re-uploaded after file permission only change

### ISSUE THREAD

**Issue by kftsehk** [aws/aws-sam-cli#2665]
### Description:
When lambda function encounter `EACCES: permission denied` error
```
    "Error: EACCES: permission denied, open '/var/task/index.js'",
    "    at Object.openSync (fs.js:462:3)",
    "    at Object.readFileSync (fs.js:364:35)",
    "    at Object.Module._extensions..js 
```

The fix is to change the file permission `chmod a+r * -R`, this can lead to a permission-only change to the lambda source code.

However, the current hashing method of lambda source code directory cannot detect permission-only change, thus will skip re-upload code artifact

The deployment will then continue using previously uploaded artifact having incorrect file permissions and still report successful. This leads to unexpected behavior that deployed lambda package have file permission different from local files.

### Steps to reproduce:
1. Create a lambda function
2. `sam deploy --guided`
3. Update _only_ file permissions of the lambda's code
4. `sam deploy --debug`

### Observed result:
```
2021-02-26 19:49:19,778 | File with same data is already exists at stackname/d3a727670c47aa78e08d049c72647445. Skipping upload
```

### Expected result:
A new package with updated permission is uploaded and deployed.

### Additional environment details (Ex: Windows, Mac, Amazon Linux etc)

1. OS: RHEL8
2. `sam --version`: 1.18.2 (pypi)
3. AWS region: ap-northeast-1

### Suggested fix

A straight forward fix is to calculate checksum _after_ zipping.
Zipping will record file permissions, hashing the zip can thus detect file permission change.

https://github.com/aws/aws-sam-cli/blob/5cbe2db23e6327f545a4e328b35bf17a3446e3d0/samcli/lib/package/utils.py#L159-L175

### Supplementary information

`dir_checksum` does not reflect permission change in files

```
>>> import hash
>>> hash.dir_checksum(".")
'3e19f2ac12731fdc3d7321b3fff0fcba'

(base) [kftse@DESKTOP-J8ITE1P utils]$ chmod og-r * 

>>> import hash
>>> hash.dir_checksum(".")
'3e19f2ac12731fdc3d7321b3fff0fcba'
```

Hash after zip seems to work

```

(base) [kftse@DESKTOP-J8ITE1P lib]$ zip utils_ur.zip utils/ -r
  adding: utils/ (stored 0%)
  ...
  adding: utils/version_checker.py (deflated 64%)
  adding: utils/__pycache__/ (stored 0%)
  adding: utils/__pycache__/hash.cpython-38.pyc (deflated 44%)
(base) [kftse@DESKTOP-J8ITE1P lib]$ chmod go+r utils/ -R
(base) [kftse@DESKTOP-J8ITE1P lib]$ zip utils_ar.zip utils/ -r
  adding: utils/ (stored 0%)
  ...
  adding: utils/version_checker.py (deflated 64%)
  adding: utils/__pycache__/ (stored 0%)
  adding: utils/__pycache__/hash.cpython-38.pyc (deflated 44%)

>>> hash.file_checksum("../utils_ar.zip")
'd1656212f988c2b2099138a1b7f7c18f'
>>> hash.file_checksum("../utils_ur.zip")
'ffc3cea3ae026947ee68dac89f3f797c'
```

**Comment by hawflau** [comment]
Hey @kftsehk , thanks for the proposal. 
Your suggested fix makes sense, but we also have a concern - this change will cause "re-deployment" to all customers because they will have different md5 even if they don't change anything in the files/folders. We will have a internal discussion about this.

Meanwhile, you can use `--force-upload` in `sam package` and `sam deploy` to force upload your package. Let me know if that helps!

**Comment by kftsehk** [comment]
Hi, @hawflau, I just realized that zipping it might not be a desirable solution depends on the situation, because any change to file metadata, e.g. `touch <filename>`, will be considered an update.

I happen to fix the problem with a combination of 
- rollback previous version
- update current version again with `--force-upload`

The main frustration is not aware of a cross-stack hidden state, the s3 deployment bucket, can be out of sync.
In the process, I had created 10s of completely new stacks to `git bisect` through a bunch of new commits, and didn't have any idea what goes wrong.
This is because all those packages I was testing were already uploaded to s3 once with wrong file permission.

The only clue was when an AWS support engineer created the function using my sample code and showed that the code has no problem, that led me to trace for hidden state in the cli, and found the hashing code in the source.
If user is aware of this problem, it is quite trivial to trigger a re-deploy by e.g. adding an empty line in code.

Imo an easier improvement can be rewrite this error message to show local directory name too, and change it from `debug` to `info`, e.g.
```
2021-02-26 19:49:19,778 | Directory ..../some-lambda with same data is already exists at stackname/d3a727670c47aa78e08d049c72647445. Skipping upload
```
